pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java

```diff
@@ -83,7 +83,16 @@ public void setUserData(final Object userData) {
 
     @Override
     public Node jjtGetChild(final int index) {
-        return (Node) children.get(index); // TODO: review if all children are Nodes
+        try {
+            return (Node) children.get(index);
+        } catch (final ClassCastException e) {
+            return new AntlrBaseNode();
+        }
+    }
+
+    @Override
+    public String getImage() {
+        return getText();
```

Comment:
In fact maybe this class should not extend from `ParserRuleContext`, otherwise we may run into compatibility problems because we couple the API of that external Antlr class to both our Java and XPath APIs. This may also cause problems like the one #1800 tries to remove (method name incompatibilities). 

 Probably, it would be best to wrap a `ParserRuleContext` instead.

Reply:
We need to extend from it because we need for Antlr parser to return us this type of nodes, we would need a bunch of adapters otherwise, we thought of that too, but we end up doing it like it is after a long talk with @jsotuyod